src/main/kotlin/org/rust/ide/annotator/RsCargoCheckAnnotator.kt

```diff
@@ -0,0 +1,179 @@
+package org.rust.ide.annotator
+
+import com.google.gson.JsonParser
+import com.intellij.lang.annotation.*
+import com.intellij.lang.annotation.Annotation
+import com.intellij.openapi.application.Result
+import com.intellij.openapi.application.WriteAction
+import com.intellij.openapi.editor.Document
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.fileEditor.FileDocumentManager
+import com.intellij.openapi.module.ModuleUtilCore
+import com.intellij.openapi.project.Project
+import com.intellij.openapi.util.ModificationTracker
+import com.intellij.openapi.util.TextRange
+import com.intellij.psi.PsiFile
+import com.intellij.psi.PsiManager
+import com.intellij.psi.util.CachedValueProvider
+import com.intellij.psi.util.CachedValuesManager
+import com.intellij.psi.util.PsiModificationTracker
+import com.intellij.util.PathUtil
+import org.apache.commons.lang.StringEscapeUtils.escapeHtml
+import org.rust.cargo.project.settings.rustSettings
+import org.rust.cargo.project.settings.toolchain
+import org.rust.cargo.toolchain.CargoMessage
+import org.rust.cargo.toolchain.CargoSpan
+import org.rust.cargo.toolchain.CargoTopMessage
+
+data class CargoCheckAnnotationInfo(val file: PsiFile, val editor: Editor)
+
+class CargoCheckAnnotationResult(commandOutput: List<String>, val project: Project)
+    : ModificationTracker by PsiManager.getInstance(project).modificationTracker {
+
+    companion object {
+        private val parser = JsonParser()
+        private val messageRegex = """\s*\{\s*"message".*""".toRegex()
+    }
+
+    val messages =
+        commandOutput
+            .filter { messageRegex.matches(it) }
+            .map { parser.parse(it) }
+            .filter { it.isJsonObject }
+            .mapNotNull { CargoTopMessage.fromJson(it.asJsonObject) }
+            .filter { it.message.message != "aborting due to previous error" }
```

Comment:
Oh, and [here's one more meta error](https://github.com/AtomBuild/atom-build-cargo/blob/736f12bb2fb0829fa6ab51f4651404fa20fb920e/lib/errors.js#L12) that should be handled the same way.